{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio import SeqIO, Entrez\n",
    "from urllib.error import HTTPError\n",
    "import pandas as pd\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "import logging\n",
    "\n",
    "\n",
    "FORMAT = \"[%(filename)s:%(lineno)s - %(funcName)20s() ] %(message)s\"\n",
    "FORMAT = \"%(filename)s:%(funcName)s():%(lineno)i: %(message)s %(levelname)s\"\n",
    "logging.basicConfig(level=logging.INFO, format=FORMAT)\n",
    "logger = logging.getLogger(__name__)\n",
    "logger.setLevel(logging.INFO)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_seq_from_FASTA(filename, as_type=\"list\"):\n",
    "    fasta_records = SeqIO.parse(open(filename),'fasta')\n",
    "    if as_type==\"dict\":\n",
    "        sequences = {}\n",
    "        for fasta_record in fasta_records:\n",
    "            sequences[fasta_record.id] = str(fasta_record.seq)\n",
    "        return sequences\n",
    "    elif as_type==\"list\":\n",
    "        sequences = []\n",
    "        for fasta_record in fasta_records:\n",
    "            sequences.append(str(fasta_record.seq))\n",
    "        return sequences\n",
    "    else:\n",
    "        raise ValueError(f\"Desired type {as_type} not supported.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Download the RNA-RNA interactions file from http://www.rnainter.org/download/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fn_db = '../data/sRNA/TableS2_E_coli.xlsx'\n",
    "data = pd.read_excel(fn_db, sheet_name='All_conditions')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqID</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>5'UTR</td>\n",
       "      <td>148</td>\n",
       "      <td>190</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>name=thrL; condition=E_coli_WT, E_coli_exo, E_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>3'UTR</td>\n",
       "      <td>255</td>\n",
       "      <td>308</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>name=thrL; condition=E_coli_WT, E_coli_exo, E_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>intergenic_UTR</td>\n",
       "      <td>255</td>\n",
       "      <td>337</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>name=thrL_thrA; condition=E_coli_pnp, E_coli_rnb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>5'UTR</td>\n",
       "      <td>544</td>\n",
       "      <td>2801</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>name=thrB; condition=E_coli_pnp, E_coli_rnb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>intergenic_UTR</td>\n",
       "      <td>2799</td>\n",
       "      <td>2801</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>name=thrA_thrB; condition=E_coli_pnp, E_coli_rnb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13994</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>5'UTR</td>\n",
       "      <td>4638329</td>\n",
       "      <td>4638355</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>name=arcA; condition=E_coli_WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13995</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>5'UTR</td>\n",
       "      <td>4638329</td>\n",
       "      <td>4638531</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>name=arcA; condition=E_coli_WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13996</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>5'UTR</td>\n",
       "      <td>4638329</td>\n",
       "      <td>4638618</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>name=arcA; condition=E_coli_WT, E_coli_exo, E_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13997</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>5'UTR</td>\n",
       "      <td>4638329</td>\n",
       "      <td>4638703</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>name=arcA; condition=E_coli_WT, E_coli_exo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13998</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>5'UTR</td>\n",
       "      <td>4638329</td>\n",
       "      <td>4638827</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>name=arcA; condition=E_coli_WT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8175 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             seqID   source         feature    start      end score strand  \\\n",
       "0      NC_000913.2  RendSeq           5'UTR      148      190     .      +   \n",
       "2      NC_000913.2  RendSeq           3'UTR      255      308     .      +   \n",
       "3      NC_000913.2  RendSeq  intergenic_UTR      255      337     .      +   \n",
       "5      NC_000913.2  RendSeq           5'UTR      544     2801     .      +   \n",
       "6      NC_000913.2  RendSeq  intergenic_UTR     2799     2801     .      +   \n",
       "...            ...      ...             ...      ...      ...   ...    ...   \n",
       "13994  NC_000913.2  RendSeq           5'UTR  4638329  4638355     .      -   \n",
       "13995  NC_000913.2  RendSeq           5'UTR  4638329  4638531     .      -   \n",
       "13996  NC_000913.2  RendSeq           5'UTR  4638329  4638618     .      -   \n",
       "13997  NC_000913.2  RendSeq           5'UTR  4638329  4638703     .      -   \n",
       "13998  NC_000913.2  RendSeq           5'UTR  4638329  4638827     .      -   \n",
       "\n",
       "      phase                                         attributes  \n",
       "0         .  name=thrL; condition=E_coli_WT, E_coli_exo, E_...  \n",
       "2         .  name=thrL; condition=E_coli_WT, E_coli_exo, E_...  \n",
       "3         .   name=thrL_thrA; condition=E_coli_pnp, E_coli_rnb  \n",
       "5         .        name=thrB; condition=E_coli_pnp, E_coli_rnb  \n",
       "6         .   name=thrA_thrB; condition=E_coli_pnp, E_coli_rnb  \n",
       "...     ...                                                ...  \n",
       "13994     .                     name=arcA; condition=E_coli_WT  \n",
       "13995     .                     name=arcA; condition=E_coli_WT  \n",
       "13996     .  name=arcA; condition=E_coli_WT, E_coli_exo, E_...  \n",
       "13997     .         name=arcA; condition=E_coli_WT, E_coli_exo  \n",
       "13998     .                     name=arcA; condition=E_coli_WT  \n",
       "\n",
       "[8175 rows x 9 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_rna = [\"3'UTR\", \"5'UTR\", 'intergenic_UTR', 'ncRNA', 'Novel_transcript', 'Novel_rRNA_or_tRNA_adjacent', 'non-coding RNA', 'pseudogene', 'putative_sRNA', 'RNase_P_RNA', 'rRNA', 'sRNA', 'SRP_RNA', 'tmRNA', 'transcript', 'tRNA']\n",
    "data['attributes'] = data['attributes'].apply(lambda x: x.replace('Name', 'name').replace('ID=', 'name='))\n",
    "data = data[data['feature'].isin(features_rna)]\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([\"5'UTR\", \"3'UTR\", 'intergenic_UTR', 'sRNA', 'non-coding RNA',\n",
       "       'Novel_transcript', 'putative_sRNA', 'transcript',\n",
       "       'Novel_rRNA_or_tRNA_adjacent', 'rRNA', 'tRNA', 'pseudogene',\n",
       "       'ncRNA', 'SRP_RNA', 'tmRNA', 'RNase_P_RNA'], dtype=object)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['attributes'] = data['attributes'].apply(lambda x: x.replace('Name', 'name'))\n",
    "data['name'] = data['attributes'].apply(lambda x: x.split('name=')[1].split(';')[0])\n",
    "data['strand num'] = data['strand'].apply(lambda x: 1 if x == '+' else 2)\n",
    "# data['sRNA_type'] = data['attributes'].apply(lambda x: x.split('sRNA_type=')[1])\n",
    "data['feature'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez, SeqIO\n",
    "import ssl\n",
    "\n",
    "# Disable SSL verification temporarily\n",
    "ssl._create_default_https_context = ssl._create_unverified_context\n",
    "\n",
    "\n",
    "def get_dna_sequence(sequence_id, start, end, strand=1):\n",
    "    \"\"\"\n",
    "    Retrieves the DNA sequence for the given sequence ID, start, and end positions.\n",
    "    \n",
    "    Parameters:\n",
    "    sequence_id (str): The ID of the DNA sequence to retrieve.\n",
    "    start (int): The starting position of the DNA sequence to retrieve.\n",
    "    end (int): The ending position of the DNA sequence to retrieve.\n",
    "    \n",
    "    Returns:\n",
    "    str: The DNA sequence for the given sequence ID, start, and end positions.\n",
    "    \"\"\"\n",
    "    # Set your email address for Entrez access\n",
    "    Entrez.email = \"olivia.gallup@gmail.com\"  # Add your email address\n",
    "    Entrez.api_key = \"8167226cf01abaa985a1c23d9b5283d22208\"  # Add your NCBI API key\n",
    "    \n",
    "    handle = Entrez.efetch(\n",
    "        db=\"nucleotide\",\n",
    "        id=sequence_id,\n",
    "        rettype=\"fasta\",\n",
    "        retmode=\"text\",\n",
    "        strand=strand,\n",
    "        seq_start=start,\n",
    "        seq_stop=end,\n",
    "    )\n",
    "    record = SeqIO.read(handle, \"fasta\")\n",
    "    \n",
    "    return str(record.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data['sequence'] = ''\n",
    "data.reset_index(drop=True, inplace=True)\n",
    "batch_size = 100\n",
    "for i0 in range(0, len(data)+batch_size, batch_size):\n",
    "    i1 = min(i0 + batch_size, len(data))\n",
    "    logging.info(f\"Processing records {i0} to {i1}...\")\n",
    "    data.loc[i0:i1, 'sequence'] = data.loc[i0:i1].apply(lambda x: get_dna_sequence(x['seqID'], x['start'], x['end'], x['strand num']), axis=1)\n",
    "    data.to_csv('../data/sRNA/TableS2_E_coli_with_seq.csv', index=False)\n",
    "\n",
    "# data['sequence'] = data.apply(lambda x: get_dna_sequence(x['seqID'], x['start'], x['end'], x['strand num']), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqID</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>name</th>\n",
       "      <th>strand num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>5'UTR</td>\n",
       "      <td>148</td>\n",
       "      <td>190</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>name=thrL; condition=E_coli_WT, E_coli_exo, E_...</td>\n",
       "      <td>thrL</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>3'UTR</td>\n",
       "      <td>255</td>\n",
       "      <td>308</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>name=thrL; condition=E_coli_WT, E_coli_exo, E_...</td>\n",
       "      <td>thrL</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>intergenic_UTR</td>\n",
       "      <td>255</td>\n",
       "      <td>337</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>name=thrL_thrA; condition=E_coli_pnp, E_coli_rnb</td>\n",
       "      <td>thrL_thrA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>5'UTR</td>\n",
       "      <td>544</td>\n",
       "      <td>2801</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>name=thrB; condition=E_coli_pnp, E_coli_rnb</td>\n",
       "      <td>thrB</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>intergenic_UTR</td>\n",
       "      <td>2799</td>\n",
       "      <td>2801</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>name=thrA_thrB; condition=E_coli_pnp, E_coli_rnb</td>\n",
       "      <td>thrA_thrB</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13994</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>5'UTR</td>\n",
       "      <td>4638329</td>\n",
       "      <td>4638355</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>name=arcA; condition=E_coli_WT</td>\n",
       "      <td>arcA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13995</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>5'UTR</td>\n",
       "      <td>4638329</td>\n",
       "      <td>4638531</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>name=arcA; condition=E_coli_WT</td>\n",
       "      <td>arcA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13996</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>5'UTR</td>\n",
       "      <td>4638329</td>\n",
       "      <td>4638618</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>name=arcA; condition=E_coli_WT, E_coli_exo, E_...</td>\n",
       "      <td>arcA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13997</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>5'UTR</td>\n",
       "      <td>4638329</td>\n",
       "      <td>4638703</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>name=arcA; condition=E_coli_WT, E_coli_exo</td>\n",
       "      <td>arcA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13998</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>5'UTR</td>\n",
       "      <td>4638329</td>\n",
       "      <td>4638827</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>name=arcA; condition=E_coli_WT</td>\n",
       "      <td>arcA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8175 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             seqID   source         feature    start      end score strand  \\\n",
       "0      NC_000913.2  RendSeq           5'UTR      148      190     .      +   \n",
       "2      NC_000913.2  RendSeq           3'UTR      255      308     .      +   \n",
       "3      NC_000913.2  RendSeq  intergenic_UTR      255      337     .      +   \n",
       "5      NC_000913.2  RendSeq           5'UTR      544     2801     .      +   \n",
       "6      NC_000913.2  RendSeq  intergenic_UTR     2799     2801     .      +   \n",
       "...            ...      ...             ...      ...      ...   ...    ...   \n",
       "13994  NC_000913.2  RendSeq           5'UTR  4638329  4638355     .      -   \n",
       "13995  NC_000913.2  RendSeq           5'UTR  4638329  4638531     .      -   \n",
       "13996  NC_000913.2  RendSeq           5'UTR  4638329  4638618     .      -   \n",
       "13997  NC_000913.2  RendSeq           5'UTR  4638329  4638703     .      -   \n",
       "13998  NC_000913.2  RendSeq           5'UTR  4638329  4638827     .      -   \n",
       "\n",
       "      phase                                         attributes       name  \\\n",
       "0         .  name=thrL; condition=E_coli_WT, E_coli_exo, E_...       thrL   \n",
       "2         .  name=thrL; condition=E_coli_WT, E_coli_exo, E_...       thrL   \n",
       "3         .   name=thrL_thrA; condition=E_coli_pnp, E_coli_rnb  thrL_thrA   \n",
       "5         .        name=thrB; condition=E_coli_pnp, E_coli_rnb       thrB   \n",
       "6         .   name=thrA_thrB; condition=E_coli_pnp, E_coli_rnb  thrA_thrB   \n",
       "...     ...                                                ...        ...   \n",
       "13994     .                     name=arcA; condition=E_coli_WT       arcA   \n",
       "13995     .                     name=arcA; condition=E_coli_WT       arcA   \n",
       "13996     .  name=arcA; condition=E_coli_WT, E_coli_exo, E_...       arcA   \n",
       "13997     .         name=arcA; condition=E_coli_WT, E_coli_exo       arcA   \n",
       "13998     .                     name=arcA; condition=E_coli_WT       arcA   \n",
       "\n",
       "       strand num  \n",
       "0               1  \n",
       "2               1  \n",
       "3               1  \n",
       "5               1  \n",
       "6               1  \n",
       "...           ...  \n",
       "13994           2  \n",
       "13995           2  \n",
       "13996           2  \n",
       "13997           2  \n",
       "13998           2  \n",
       "\n",
       "[8175 rows x 11 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.to_csv('../data/sRNA/TableS2_E_coli_with_seq.csv', index=False)\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
