{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio import SeqIO, Entrez\n",
    "from urllib.error import HTTPError\n",
    "import pandas as pd\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_seq_from_FASTA(filename, as_type=\"list\"):\n",
    "    fasta_records = SeqIO.parse(open(filename),'fasta')\n",
    "    if as_type==\"dict\":\n",
    "        sequences = {}\n",
    "        for fasta_record in fasta_records:\n",
    "            sequences[fasta_record.id] = str(fasta_record.seq)\n",
    "        return sequences\n",
    "    elif as_type==\"list\":\n",
    "        sequences = []\n",
    "        for fasta_record in fasta_records:\n",
    "            sequences.append(str(fasta_record.seq))\n",
    "        return sequences\n",
    "    else:\n",
    "        raise ValueError(f\"Desired type {as_type} not supported.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Download the RNA-RNA interactions file from http://www.rnainter.org/download/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "fn_db = '../data/sRNA/TableS2_E_coli.xlsx'\n",
    "data = pd.read_excel(fn_db, sheet_name='sRNAs')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqID</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>name</th>\n",
       "      <th>sRNA_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>75517</td>\n",
       "      <td>75608</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>name=sroA; E_coli_WT; start_dist=-0.0; stop_di...</td>\n",
       "      <td>sroA</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>92418</td>\n",
       "      <td>92658</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>name=ftsI_ftsO; E_coli_WT; start_dist=-67.0; s...</td>\n",
       "      <td>ftsI_ftsO</td>\n",
       "      <td>intragenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>92485</td>\n",
       "      <td>92658</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>name=ftsI_ftsO; E_coli_WT; start_dist=0.0; sto...</td>\n",
       "      <td>ftsI_ftsO</td>\n",
       "      <td>intragenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>344476</td>\n",
       "      <td>344638</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>name=xtpA_yahM; E_coli_WT; start_dist=0.0; sto...</td>\n",
       "      <td>xtpA_yahM</td>\n",
       "      <td>5'UTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>506428</td>\n",
       "      <td>506511</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>name=chiX; E_coli_WT; start_dist=0.0; stop_dis...</td>\n",
       "      <td>chiX</td>\n",
       "      <td>independent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>4188346</td>\n",
       "      <td>4188510</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>name=sroH; E_coli_WT; start_dist=-0.0; stop_di...</td>\n",
       "      <td>sroH</td>\n",
       "      <td>independent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>4261197</td>\n",
       "      <td>4261269</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>name=pspH; E_coli_WT; start_dist=35.0; stop_di...</td>\n",
       "      <td>pspH</td>\n",
       "      <td>independent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>4525999</td>\n",
       "      <td>4526135</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>name=ryjB; E_coli_WT; start_dist=-1.0; stop_di...</td>\n",
       "      <td>ryjB</td>\n",
       "      <td>independent_w_isoform</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>4526044</td>\n",
       "      <td>4526135</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>name=ryjB; E_coli_WT; start_dist=44.0; stop_di...</td>\n",
       "      <td>ryjB</td>\n",
       "      <td>independent_w_isoform</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>NC_000913.2</td>\n",
       "      <td>RendSeq</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>4577858</td>\n",
       "      <td>4577935</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>name=symR; E_coli_WT; start_dist=0.0; stop_dis...</td>\n",
       "      <td>symR</td>\n",
       "      <td>independent</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>91 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          seqID   source feature    start      end score strand phase  \\\n",
       "0   NC_000913.2  RendSeq    sRNA    75517    75608     .      -     .   \n",
       "1   NC_000913.2  RendSeq    sRNA    92418    92658     .      +     .   \n",
       "2   NC_000913.2  RendSeq    sRNA    92485    92658     .      +     .   \n",
       "3   NC_000913.2  RendSeq    sRNA   344476   344638     .      +     .   \n",
       "4   NC_000913.2  RendSeq    sRNA   506428   506511     .      +     .   \n",
       "..          ...      ...     ...      ...      ...   ...    ...   ...   \n",
       "86  NC_000913.2  RendSeq    sRNA  4188346  4188510     .      -     .   \n",
       "87  NC_000913.2  RendSeq    sRNA  4261197  4261269     .      +     .   \n",
       "88  NC_000913.2  RendSeq    sRNA  4525999  4526135     .      +     .   \n",
       "89  NC_000913.2  RendSeq    sRNA  4526044  4526135     .      +     .   \n",
       "90  NC_000913.2  RendSeq    sRNA  4577858  4577935     .      +     .   \n",
       "\n",
       "                                           attributes       name  \\\n",
       "0   name=sroA; E_coli_WT; start_dist=-0.0; stop_di...       sroA   \n",
       "1   name=ftsI_ftsO; E_coli_WT; start_dist=-67.0; s...  ftsI_ftsO   \n",
       "2   name=ftsI_ftsO; E_coli_WT; start_dist=0.0; sto...  ftsI_ftsO   \n",
       "3   name=xtpA_yahM; E_coli_WT; start_dist=0.0; sto...  xtpA_yahM   \n",
       "4   name=chiX; E_coli_WT; start_dist=0.0; stop_dis...       chiX   \n",
       "..                                                ...        ...   \n",
       "86  name=sroH; E_coli_WT; start_dist=-0.0; stop_di...       sroH   \n",
       "87  name=pspH; E_coli_WT; start_dist=35.0; stop_di...       pspH   \n",
       "88  name=ryjB; E_coli_WT; start_dist=-1.0; stop_di...       ryjB   \n",
       "89  name=ryjB; E_coli_WT; start_dist=44.0; stop_di...       ryjB   \n",
       "90  name=symR; E_coli_WT; start_dist=0.0; stop_dis...       symR   \n",
       "\n",
       "                sRNA_type  \n",
       "0              intergenic  \n",
       "1              intragenic  \n",
       "2              intragenic  \n",
       "3                   5'UTR  \n",
       "4             independent  \n",
       "..                    ...  \n",
       "86            independent  \n",
       "87            independent  \n",
       "88  independent_w_isoform  \n",
       "89  independent_w_isoform  \n",
       "90            independent  \n",
       "\n",
       "[91 rows x 11 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['name'] = data['attributes'].apply(lambda x: x.split('name=')[1].split(';')[0])\n",
    "data['sRNA_type'] = data['attributes'].apply(lambda x: x.split('sRNA_type=')[1])\n",
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "Entrez.email = \"olivia.gallup@gmail.com\"  # Add your email address\n",
    "Entrez.api_key = \"8167226cf01abaa985a1c23d9b5283d22208\"  # Add your NCBI API key\n",
    "\n",
    "\n",
    "def get_seq_from_genome(gene_id):\n",
    "    try:\n",
    "        handle = Entrez.efetch(db=\"gene\", id=gene_id, rettype=\"fasta\", retmode=\"xml\")\n",
    "        record = Entrez.read(handle)\n",
    "        nuc_id = record[0]['Entrezgene_locus'][0]['Gene-commentary_accession'] + '.' + record[0]['Entrezgene_locus'][0]['Gene-commentary_version']\n",
    "\n",
    "        dna_start = record[0]['Entrezgene_locus'][0]['Gene-commentary_seqs'][0]['Seq-loc_int']['Seq-interval']['Seq-interval_from']\n",
    "        dna_end = record[0]['Entrezgene_locus'][0]['Gene-commentary_seqs'][0]['Seq-loc_int']['Seq-interval']['Seq-interval_to']\n",
    "\n",
    "        search_handle = Entrez.efetch(db=\"nuccore\", id=nuc_id, retmode=\"XML\", rettype='fasta_cds_na', strand=\"complement\", seq_start=int(dna_start), seq_stop=int(dna_end))\n",
    "        # dna_sequence = SeqIO.read(search_handle, \"fasta\").seq\n",
    "        r = search_handle.read()\n",
    "        if r:\n",
    "            dna_sequence = ''.join(r.split('>')[1].split('\\n')[1:])\n",
    "        else:\n",
    "            dna_sequence = ''\n",
    "        return dna_sequence\n",
    "    except HTTPError:\n",
    "        return ''\n",
    "\n",
    "\n",
    "def get_seq_info(gene_id):\n",
    "    try:\n",
    "        handle = Entrez.efetch(db=\"nucleotide\", id=gene_id, rettype=\"fasta\", retmode=\"text\")\n",
    "        fast = handle.read()\n",
    "        seq = ''.join(fast.split('\\n')[1:])\n",
    "        if not seq or ('N' in seq):\n",
    "            seq = get_seq_from_genome(gene_id)\n",
    "        return seq\n",
    "    except HTTPError:\n",
    "        seq = get_seq_from_genome(gene_id)\n",
    "        return seq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = {k: get_seq_info(k) for k in list(data['Raw_ID1'].str.replace('NCBI:', '').apply(str).unique()) + list(data['Raw_ID2'].str.replace('NCBI:', '').apply(str).unique())}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# if ('Sequence1' not in data.columns) or ('Sequence2' not in data.columns):\n",
    "# data['Sequence1'] = data['Raw_ID1'].str.replace('NCBI:', '').apply(str).apply(get_fasta)\n",
    "d = {k: get_seq_info(k) for k in list(data['Raw_ID1'].str.replace('NCBI:', '').apply(str).unique()) + list(data['Raw_ID2'].str.replace('NCBI:', '').apply(str).unique())}\n",
    "\n",
    "for i, l in enumerate(zip(data['Raw_ID1'].str.replace('NCBI:', '').apply(str), data['Raw_ID2'].str.replace('NCBI:', '').apply(str))):\n",
    "    data[f'Sequence{i+1}'] = list(map(lambda x: d[x], l))\n",
    "    # data[f'Gene_Start{i+1}'] = list(map(lambda x: d[x][1], l))\n",
    "    # data[f'Gene_End{i+1}'] = list(map(lambda x: d[x][2], l))\n",
    "# data['Sequence1'] = data['Raw_ID1'].str.replace('NCBI:', '').apply(str).apply(get_fasta)\n",
    "\n",
    "print(sum((data['Sequence1'] != '') & (data['Sequence2'] != '')))\n",
    "data.to_csv(fn)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>level_0</th>\n",
       "      <th>index</th>\n",
       "      <th>RNAInterID</th>\n",
       "      <th>Interactor1.Symbol</th>\n",
       "      <th>Category1</th>\n",
       "      <th>Species1</th>\n",
       "      <th>Interactor2.Symbol</th>\n",
       "      <th>Category2</th>\n",
       "      <th>Species2</th>\n",
       "      <th>Raw_ID1</th>\n",
       "      <th>Raw_ID2</th>\n",
       "      <th>score</th>\n",
       "      <th>strong</th>\n",
       "      <th>weak</th>\n",
       "      <th>predict</th>\n",
       "      <th>Sequence1</th>\n",
       "      <th>Sequence2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [level_0, index, RNAInterID, Interactor1.Symbol, Category1, Species1, Interactor2.Symbol, Category2, Species2, Raw_ID1, Raw_ID2, score, strong, weak, predict, Sequence1, Sequence2]\n",
       "Index: []"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data[data['Sequence1'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
